native/lib.rs

```diff
@@ -58,29 +60,34 @@ fn put_5<'a>(
     max_size: usize,
     min_q1_size: usize,
 ) -> Token {
-    let mut caches = CACHES.lock().unwrap();
     let cache_name_bin = HBin::from_atom(env, cache_name);
-    let cache = match caches.get_mut(&cache_name_bin) {
-        Some(cache) => cache,
-        None => {
-            let recent_ratio = min_q1_size as f64 / max_size as f64;
-            let inner = TwoQueueCacheBuilder::new(max_size)
-                .set_recent_ratio(recent_ratio)
-                .finalize()
-                .map_err(|e| {
-                    eprintln!(
-                        "MinQ1Size/MaxSize ({}/{}) ratio doesn't appear to make sense ({})",
-                        min_q1_size, max_size, e
-                    )
-                })
-                .expect("MinQ1Size/MaxSize ratio doesn't appear to make sense");
-            let cache = TermCache::new(inner);
-            caches.insert(HBin::from_atom(env, cache_name), cache);
-            caches
-                .get_mut(&cache_name_bin)
-                .expect("cache was just inserted. It must exist")
+    let cache = {
+        let caches = CACHES.read().unwrap();
+        match caches.get(&cache_name_bin) {
+            Some(cache) => cache.clone(),
+            None => {
+                let recent_ratio = min_q1_size as f64 / max_size as f64;
+                let inner = TwoQueueCacheBuilder::new(max_size)
+                    .set_recent_ratio(recent_ratio)
+                    .finalize()
+                    .map_err(|e| {
+                        eprintln!(
+                            "MinQ1Size/MaxSize ({}/{}) ratio doesn't appear to make sense ({})",
+                            min_q1_size, max_size, e
+                        )
+                    })
+                    .expect("MinQ1Size/MaxSize ratio doesn't appear to make sense");
+                let cache = Arc::new(RwLock::new(TermCache::new(inner)));
+                // we need to manually drop caches to prevent a
+                // deadlock when creating the upcoming write lock.
+                drop(caches);
+                let mut caches = CACHES.write().unwrap();
+                caches.insert(HBin::from_atom(env, cache_name), cache.clone());
```

Comment:
I think there's the possibility of a race condition here where the `put` first discovers there is no cache under the read lock, drops the lock, and another thread in `put` or `create` adds the cache. The `insert` here that happens next erases the previous cache, possibly causing the loss of a cached item. Since the cache is an optimization, this isn't a correctness problem but it could be a minor performance problem.